# Machine Learning - Final project
### Empathy Prediction using the [Young people survey](https://www.kaggle.com/miroslavsabo/young-people-survey/) dataset

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set(style="white", color_codes=True)
sns.set_context(rc={"font.family":'sans',"font.size":24,"axes.titlesize":24,"axes.labelsize":24})  
pd.options.display.max_columns = 150
pd.options.display.max_rows = 150

#### Loading datasets

In [89]:
df = pd.read_csv('./input/responses.csv')
columns_desc = pd.read_csv('./input/columns.csv')

In [90]:
df.shape

(1010, 150)

#### Defining way to access description easily

In [70]:
all_columns =list(df)

desc = {}
for col_name in all_columns:
    desc[col_name] = columns_desc.loc[columns_desc['short'] == col_name].iloc[0][0]
    
def print_desc(column_name):
    print(column_name+" --> "+desc[column_name])

Example:

In [71]:
for col in all_columns[:5]:
    print_desc(col)

Music --> I enjoy listening to music.
Slow songs or fast songs --> I prefer.
Dance --> Dance, Disco, Funk
Folk --> Folk music
Country --> Country


## Data Exploration and preprocessing

### Missing values

Let's start by looking at how many rows are missing a class label, I will delete these rows since it makes no sense to impute them, train or test using them

In [86]:
df['Empathy'].isna().sum()

5

In [91]:
df.dropna(subset=['Empathy'], inplace=True)

In [92]:
df.shape

(1005, 150)

Let's now look at the percentage of missing values in all columns in desc order

In [94]:
df_na = (df.isnull().sum() / len(df)) * 100
df_na = df_na.drop(df_na[df_na == 0].index).sort_values(ascending=False)
missing_data = pd.DataFrame({'Missing Ratio': df_na})
missing_data.head(150).T

Height   Weight  Passive sport  Chemistry  Geography  Theatre  \
Missing Ratio  1.99005  1.99005       1.492537   0.995025   0.895522  0.79602   

               Documentary  Smoking   Latino     Punk  Criminal damage  \
Missing Ratio      0.79602  0.79602  0.79602  0.79602         0.696517   

               Compassion to animals  Final judgement  Reggae, Ska  Gardening  \
Missing Ratio               0.696517         0.696517     0.696517   0.696517   

               Alternative  Rock n roll  Techno, Trance       Age  \
Missing Ratio     0.696517     0.696517        0.696517  0.696517   

               Classical music  Swing, Jazz    Movies   Biology   Reading  \
Missing Ratio         0.696517     0.597015  0.597015  0.597015  0.597015   

                 Giving        PC    Gender  Number of siblings  \
Missing Ratio  0.597015  0.597015  0.597015            0.597015   

               Responding to a serious letter  Daily events  \
Missing Ratio                        0.597015      0.597015   

               Science and technology  Art exhibitions  Friends versus money  \
Missing Ratio                0.597015         0.597015              0.597015   

                Writing  Countryside, outdoors      Rock      Folk  \
Missing Ratio  0.597015               0.597015  0.597015  0.497512   

               Self-criticism   Spiders   Country  Economy Management  \
Missing Ratio        0.497512  0.497512  0.497512            0.497512   

               Energy levels  Workaholism  Psychology  Foreign languages  \
Missing Ratio       0.497512     0.497512    0.497512           0.497512   

               Prioritising workload  Getting up  Medicine   Alcohol  \
Missing Ratio               0.497512    0.497512  0.497512  0.497512   

               Socializing  Village - town     Pets  Active sport  \
Missing Ratio     0.497512         0.39801  0.39801       0.39801   

               Reliability  Loss of interest  Fun with friends     Cars  \
Missing Ratio      0.39801           0.39801           0.39801  0.39801   

               House - block of flats  Funniness  Decision making    Dance  \
Missing Ratio                 0.39801    0.39801          0.39801  0.39801   

               Hiphop, Rap  Questionnaires or polls  Finding lost valuables  \
Missing Ratio      0.39801                  0.39801                 0.39801   

               Personality  Small - big dogs  Happiness in life  \
Missing Ratio      0.39801           0.39801            0.39801   

               Getting angry  Children  Mood swings  Western  Internet  \
Missing Ratio        0.39801   0.39801      0.39801  0.39801   0.39801   

               Cheating in school  Hypochondria  Judgment calls  Romantic  \
Missing Ratio             0.39801       0.39801         0.39801  0.298507   

                 Comedy  Fantasy/Fairy tales  Animated  Religion   Physics  \
Missing Ratio  0.298507             0.298507  0.298507  0.298507  0.298507   

                Dancing  Metal or Hardrock       Pop  Mathematics     Music  \
Missing Ratio  0.298507           0.298507  0.298507     0.298507  0.298507   

               Appearence and gestures  Spending on looks  Life struggles  \
Missing Ratio                 0.298507           0.298507        0.298507   

               Interests or hobbies   Charity  Finances  Elections  \
Missing Ratio              0.298507  0.298507  0.298507   0.298507   

               Thinking ahead  Writing notes  Healthy eating  \
Missing Ratio        0.298507       0.298507        0.298507   

               Entertainment spending      Rats   Heights  Left - right handed  
Missing Ratio                0.298507  0.298507  0.298507             0.298507

The percentage of missing values is not a lot, we can easily impute them without introducing too much bias by using the mode

In [95]:
df[all_columns]=df[all_columns].fillna(df.mode().iloc[0])

In [97]:
df.isnull().any().any()

False

There are no missing values now

### Data types

In [17]:
df['Music'] = df['Music'].astype(int)

['a', 'b', 'c', 'd']

In [105]:
df.dtypes[:10]

Music                       float64
Slow songs or fast songs    float64
Dance                       float64
Folk                        float64
Country                     float64
Classical music             float64
Musical                     float64
Pop                         float64
Rock                        float64
Metal or Hardrock           float64
dtype: object

In [109]:
float_cols = df.dtypes == np.float64

In [115]:
fake_float = list(df.loc[:, float_cols])

In [116]:
len(fake_float)

134

134 columns are float but should be int since the decimal value is always 0, let's convert them

In [117]:
df[fake_float] = df[fake_float].astype(int)

### One hot encoding
Since the algorithms that I have in mind do not handle categorical values I will need to perform one hot encoding or encode values in numbers for ordinal data

Let's look at categorical columns and decide which ones can be ordinal

In [121]:
categorical = df.dtypes == np.object

In [123]:
df.loc[:, categorical].T

0                        1     \
Smoking                            never smoked             never smoked   
Alcohol                             drink a lot              drink a lot   
Punctuality                 i am always on time         i am often early   
Lying                                     never                sometimes   
Internet usage                  few hours a day          few hours a day   
Gender                                   female                   female   
Left - right handed                right handed             right handed   
Education               college/bachelor degree  college/bachelor degree   
Only child                                   no                       no   
Village - town                          village                     city   
House - block of flats           block of flats           block of flats   

                                           2     \
Smoking                           tried smoking   
Alcohol                             drink a lot   
Punctuality             i am often running late   
Lying                                 sometimes   
Internet usage                  few hours a day   
Gender                                   female   
Left - right handed                right handed   
Education                      secondary school   
Only child                                   no   
Village - town                             city   
House - block of flats           block of flats   

                                                 3                      4     \
Smoking                                 former smoker          tried smoking   
Alcohol                                   drink a lot         social drinker   
Punctuality                          i am often early    i am always on time   
Lying                   only to avoid hurting someone  everytime it suits me   
Internet usage                        most of the day        few hours a day   
Gender                                         female                 female   
Left - right handed                      right handed           right handed   
Education                     college/bachelor degree       secondary school   
Only child                                        yes                     no   
Village - town                                   city                village   
House - block of flats                 house/bungalow         house/bungalow   

                                                 5     \
Smoking                                  never smoked   
Alcohol                                         never   
Punctuality                          i am often early   
Lying                   only to avoid hurting someone   
Internet usage                        few hours a day   
Gender                                           male   
Left - right handed                      right handed   
Education                            secondary school   
Only child                                         no   
Village - town                                   city   
House - block of flats                 block of flats   

                                           6                        7     \
Smoking                           tried smoking           current smoker   
Alcohol                          social drinker              drink a lot   
Punctuality                    i am often early      i am always on time   
Lying                                     never                sometimes   
Internet usage          less than an hour a day          few hours a day   
Gender                                   female                     male   
Left - right handed                right handed             right handed   
Education                      secondary school  college/bachelor degree   
Only child                                   no                       no   
Village - town                          village                     city   
House - block of flats           ho

In [128]:
categorical_cols = list(df.loc[:, categorical])

In [130]:
for col in categorical_cols:
    print_desc(col)

Smoking --> Smoking habits
Alcohol --> Drinking
Punctuality --> Timekeeping.
Lying --> Do you lie to others?
Internet usage --> How much time do you spend online?
Gender --> Gender
Left - right handed --> I am
Education --> Highest education achieved
Only child --> I am the only child
Village - town --> I spent most of my childhood in a
House - block of flats --> I lived most of my childhood in a


There are a lot of ordinal columns, which is better because we will not occur in a feature explosion using OHE, I think the ordinal values could be:

* Alcohol 
* Punctuality
* Internet usage
* Education

Some other that only have True/False or Yes/No or in general 2 possible values can be encoded using the same attribute and an integer value (0 or 1), these features are:

* Gender
* Left - right handed
* Only child
* Village - town
* House - block of flats

The remaining categorical attributes are the ones that I think are the most important to predict empathy or that are not really ordinal and I would not want to introduce a bias by considering them as ordinal, these attributes are:

